setup.sh

```diff
@@ -1,76 +1,175 @@
-git clone https://github.com/zsh-users/zsh-autosuggestions ${ZSH_CUSTOM:-~/.oh-my-zsh/custom}/plugins/zsh-autosuggestions
-git clone https://github.com/zsh-users/zsh-syntax-highlighting.git ${ZSH_CUSTOM:-~/.oh-my-zsh/custom}/plugins/zsh-syntax-highlighting
-
-if [ -f "${HOME}/.zshrc" ]; then
-  ZSHRC_BACKUP="${HOME}/.zshrc.backup"
-  while [ -f "${ZSHRC_BACKUP}" ]; do
-    ZSHRC_BACKUP="${ZSHRC_BACKUP}.backup"
-  done
-  ZSHRC_BACKUP_REL=$(realpath --relative-to="${HOME}" "${ZSHRC_BACKUP}")
-  echo "~/.zshrc exists, moving it to ~/${ZSHRC_BACKUP_REL}"
-  mv ${HOME}/.zshrc ${ZSHRC_BACKUP}
-fi
+# Usage: see -h / usage()
 
-NO_ZSHRC=false
-if [ -f "${HOME}/.zshrc" ]; then
-  NO_ZSHRC=true
-else
-  curl -fsSL https://raw.githubusercontent.com/STAOJ/sta-setup/master/.zshrc > ${HOME}/.zshrc
-fi
+#######################################
+# Adds the specified string to the shell-rc files if it doesn't already exist.
+# This adds the specified string to both .zshrc and .bashrc if they exist.
+# Globals:
+#   HOME
+# Arguments:
+#   $1: the string to be added to .bashrc.
+#######################################
 
-if [ -f "${HOME}/.p10k.zsh" ]; then
-  P10K_BACKUP="${HOME}/.p10k.zsh.backup"
-  while [ -f "${P10K_BACKUP}" ]; do
-    P10K_BACKUP="${P10K_BACKUP}.backup"
-  done
-  P10K_BACKUP_REL=$(realpath --relative-to="${HOME}" "${P10K_BACKUP}")
-  echo "~/.p10k.zsh exists, moving it to ~/${P10K_BACKUP_REL}"
-  mv ${HOME}/.p10k.zsh ${P10K_BACKUP}
-fi
+add_to_rc () {
+    for shell_file in "$HOME/.bashrc" "$HOME/.zshrc"; do
+        if  [ -f "$shell_file" ]; then
+            append_if_not_present "$1" "$shell_file"
+        fi
+    done
+}
 
-NO_P10K=false
-if [ -f "${HOME}/.p10k.zsh" ]; then
-  NO_P10K=true 
-else
-  curl -fsSL https://raw.githubusercontent.com/STAOJ/sta-setup/master/.p10k.zsh > ${HOME}/.p10k.zsh
-fi
+#######################################
+# Adds the given string to the given file if it doesn't already exist.
+# Arguments:
+#   $1: the string to be added to the file.
+#   $2: the path to the file.
+#######################################
 
-if [ "$NO_SHELL" = true ]; then
-  echo "Install successful, you don't seem to be using either bash or zsh so we can't figure out how to change your shell to zsh. Using zsh is not necessary, but it has some nice things installed like syntax highlighting and auto completion"
-fi
+append_if_not_present () {
+grep -qx "$1" "$2" || echo "$1" >> "$2"
+}
 
-if [ "$NO_ZSHRC" = true ]; then
-  echo "~/.zshrc already exists and we are unable to create an backup of it. So the recommended ~/.zshrc was not installed. To install it, consider removing your ~/.zshrc and run \n    curl -fsSL https://raw.githubusercontent.com/STAOJ/sta-setup/master/.zshrc > ${HOME}/.zshrc"
-fi
+#######################################
+# Installs brew under the local user from git.
+# Globals:
+#   HOME
+# Arguments:
+#   None
+#######################################
+
+stasetup::install_brew () {
+    set -e 
```

Comment:
I guess you want to exit on the first error because that indicates that homebrew is installed, but I don't think this is the best way to do it, because it will also kill the calling code (anything after this function will not be run). Perhaps you should add some test at the start to test if homebrew is already installed, and return if so, or add `&&` to the end of every line so that the sequence stops at the earliest error.

Reply:
All the code runs without error whether homebrew is already installed or not (as ensured in 75eaee5 ) - this just detects errors in the installation process.  

I agree that stopping the calling shell might not be desirable in all cases - I agree that replacing this with proper error handling ( ` dothis || return 1` ) might be better...

(As a side note - using `set -e` seems to be fine in some cases but not others, as described in https://stackoverflow.com/a/69808. It could be that setting `set -e` on a global level for this script is a good idea, however this would require going back through the zsh code to make sure that it doesn't fail if zsh is already installed as i did with the brew code.... )

Reply:
@DE0CH There is no good answer to this it seems....

I can use a trap inside the function to catch errors and return 1 early, but this no longer works if you do `function || echo "function has failed"`.  (the same applies to `set -e` too.)

I cant see a way to make the function return on error robustly while keeping the calling shell open.

I think the best way to do this is to leave this to global error handling, and do nothing inside the functions.
 That is, put something like this at the top of the file:
```bash
set -E # pass errors down from subshells / functions
trap 'echo "an error has occurred running $FUNCNAME at line $LINENO; exit 1" 
```

This however, would mean that the zsh code would need to be fixed so that it doesn't throw an error if zsh is already installed / a directory already exists / etc... Because of this, I've removed error handling from this PR for now.

Reply:
beb0166

Reply:
Thanks for spending the time to investigate this. I see. It's probably going to take some work to implement this, so I think it's fine if error handling is not perfect at this moment. We can open an issue to keep track of this and improve this in a future PR. Thanks again for your dedication!! :D

Reply:
https://github.com/CP-STA/sta-setup/issues/2#issue-1290459909